File: ver-mac-x64/src/budg/budg.cpp
#include <any>
#include <vector>

#include "budg.h"
#include "budgAux.h"
#include "budgEffect.h"
#include "ignore.kd.h"
#include "VM.h"

#define CB(code) std::make_any<std::function<void(BudgetContext)>>([&](BudgetContext c) { code })
#define VM VM::singleton()

BudgComponent::BudgComponent() {
    std::vector<std::any> oneliners = {
        F.didClickCopy, CB( budgCopyResult(c.result()); ),
        F.didClickPasteMorningBalance, CB( budgPasteMorningBalance(); ),
        F.didClickPasteSpent, CB( budgPasteSpent(); ),
        F.morningBalance, CB( VM.setMorningBalance(c.morningBalance()); ),
        F.result, CB( VM.setResult(c.result()); ),
        F.spent, CB( VM.setSpent(c.spent()); ),
    };
    BudgEffectRegistry::registerOneliners(KT.budgCtrl(), oneliners);
}

void BudgComponent::setup() {
    // Defaults
    budgSet(F.reportedDate, budgReportedDate());
    budgSet(F.reportedWeekday, budgReportedWeekday());
    budgSet(F.didSetup, true);
}

File: ver-ios/src/budg/budg.swift
import gb

private typealias BC = BudgContext

class BudgComponent {
    init() {
        var vm = { VM.singleton! }

        let effects: [Any] = [
            F.didClickCopy, { (c: BC) in budgCopyResult(c.result) },
            F.didClickPasteMorningBalance, { (c: BC) in budgPasteMorningBalance() },
            F.didClickPasteSpent, { (c: BC) in budgPasteSpent() },
            F.morningBalance, { (c: BC) in vm().morningBalance = c.morningBalance },
            F.spent, { (c: BC) in vm().spent = c.spent },
            F.result, { (c: BC) in vm().result = c.result },
        ]
        var r: BC? = registerOneliners(budgCtrl(), effects)
    }

    func setup() {
        // Defaults
        budgSet(F.reportedDate, budgReportedDate())
        budgSet(F.reportedWeekday, budgReportedWeekday())
        budgSet(F.didSetup, true)
    }
}
